src/PatchOrchestrationApplication/NodeAgentSFUtility/src/Helpers/HealthManagerHelper.cs

```diff
@@ -49,23 +54,92 @@ class HealthManagerHelper
+                sendOptions.Immediate = true;
+                TimeoutProcessWithRetry(() =>
+                {
+                    fabricClient.HealthManager.ReportHealth(healthReport, sendOptions);
+                }, condition,
+                serviceNameSuffix, timeout
+                );
+
+                Task.Delay(TimeSpan.FromSeconds(2)).GetAwaiter().GetResult();
                 return NodeAgentSfUtilityExitCodes.Success;
             }
             catch (Exception e)
             {
                 ServiceEventSource.Current.ErrorMessage(
-                    String.Format("HealthManagerHelper.PostNodeHealthReport failed. Exception details {0}", e));
+                    String.Format("HealthManagerHelper.PostNodeHealthReport for Service {0} failed. Exception details {1}", serviceNameSuffix, e));
                 if (e is FabricTransientException)
                 {
                     return NodeAgentSfUtilityExitCodes.RetryableException;
                 }
+                else if(e is TimeoutException)
+                {
+                    return NodeAgentSfUtilityExitCodes.TimeoutException;
+                }
                 else
                 {
                     return NodeAgentSfUtilityExitCodes.Failure;
                 }
             }
         }
+
+        /// <summary>
+        /// Retry utility to check for a condition to run the function specified in params.
+        /// </summary>
+        /// <param name="condition">Condition to check before running the function passed./</param>
+        /// <param name="process">function to run if condition passes within configured timeout.</param>
+        /// <param name="timeout">Timeout configured for this operation.</param>
+        internal static void TimeoutProcessWithRetry(Action process, Func<string, bool> condition, string s, TimeSpan timeout)
+        {
+            Stopwatch stopwatch = new Stopwatch();
+            stopwatch.Start();
+            int count = 0;
+            while(stopwatch.Elapsed < timeout)
+            {
+                if(condition(s))
+                {
+                    process();
+                    return;
+                }
+                count++;
+                Thread.Sleep(TimeSpan.FromSeconds(1 * count));
+            }
+            stopwatch.Reset();
+            throw new TimeoutException(string.Format("Timeout occurred while trying to run the process {0}", process.GetType().Name.ToString()));
+        }
+
+        /// <summary>
+        /// This method checks if the service with specified suffix is up or not.
+        /// </summary>
+        /// <param name="fabricClient">Fabric client./</param>
+        /// <param name="applicationName">Application name of the service</param>
+        /// <param name="timeout">Suffix of the service to check.</param>
+        internal static async Task<bool> CheckIfServiceIsUp(FabricClient fabricClient, Uri applicationName, string serviceNameSuffix)
```

Comment:
>CheckIfServiceIsUp [](start = 41, length = 18)

The method checks if the service exists or not. Exists and up are two different things. What is the failure scenario you are trying to handle here?